python/ray/util/ml_utils/checkpoint_manager.py

```diff
@@ -0,0 +1,315 @@
+import gc
+import heapq
+import logging
+import numbers
+import os
+import shutil
+
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Optional, Dict, Union, Callable, Tuple, List, Any
+
+import ray
+from ray.tune.result import NODE_IP, TRAINING_ITERATION
+from ray.util import PublicAPI
+from ray.util.ml_utils.util import is_nan
+
+MAX = "max"
+MIN = "min"
+
+logger = logging.getLogger(__name__)
+
+
+class _TrackedCheckpoint:
+    MEMORY = "memory"
+    PERSISTENT = "persistent"
+
+    def __init__(
+        self,
+        checkpoint_dir_or_data: Optional[Union[str, Path, Dict, ray.ObjectRef]],
+        storage_mode: str,
+        checkpoint_id: Optional[int] = None,
+        result: Optional[Dict] = None,
+        node_ip: Optional[str] = None,
+    ):
+        self.checkpoint_dir_or_data = checkpoint_dir_or_data
+        self.checkpoint_id = checkpoint_id
```

Comment:
maybe just `dir_or_data` and `id`? We don't need to specify those concern a checkpoint when we are already in a checkpoint object

Reply:
Great point, updated